### PR TITLE
uwsm: prefer user env binaries at runtime

### LINKED ISSUE #358805: uwsm: starting apps with uwsm have random paths added
## Describe the bug

After enabling uwsm programs that are started with `uwsm app -- program` have some nix paths added. This leads to running bash instead of bashInteractive in a terminal.

## Steps To Reproduce

1. Enable uwsm
2. Open terminal with `uwsm app --foot`
3. Try to run `bash`
4. See `bash: shopt: progcomp: invalid shell option name` and weird rendering

## Expected behavior

No new paths added.

## Screenshots

With uwsm app -- foot:
image
image

Should be:
image

## Metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.11.8, NixOS, 25.05 (Warbler), 25.05.20241115.5e4fbfb`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.91.1
System type: x86_64-linux
Additional system types: i686-linux, x86_64-v1-linux, x86_64-v2-linux, x86_64-v3-linux
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/karun/.config/nix/nix.conf:/nix/store/gqpgq34fbzmxaak73j7a1jkph5v3chav-stylix-kde-config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/karun/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/karun/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/karun/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/kqpjnkkpv1drbnvrqbhnyd2ixy9dymzb-lix-2.91.1/share`
 - channels(root): `"home-manager, nixos-24.05"`
 - nixpkgs: `/nix/store/hfz1qqd0z8amlgn8qwich1dvkmldik36-source`

## Notify maintainers

@JohnRTitor @kai-tub 

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by JohnRTitor**
We are wrapping uwsm with bash and other packages, else it won't work because it won't find the dependencies at runtime (it's just a python script). However this also leaks the `PATH` to subprocesses/apps that are launched using UWSM. Leading to issues like https://github.com/NixOS/nixpkgs/issues/358805.

In this PR, we just change the path order of these runtime dependencies, so these are treated like fallbacks. Ideally we should not expect anything from user environment, and the wrapping approach isn't the best. Not like we have any other choice.

CC @Scrumplex @Atemu fellow UWSM users, they might have an idea to fix this.
CC @Vladimir-csp, UWSM dev

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/uw/uwsm/package.nix [hunk 1]
```diff
@@ -66,7 +68,7 @@ stdenv.mkDerivation (finalAttrs: {
   postInstall =
     let
       wrapperArgs = ''
-        --prefix PATH : "${lib.makeBinPath finalAttrs.propagatedBuildInputs}"
+        --suffix PATH : "${lib.makeBinPath finalAttrs.propagatedBuildInputs}"
```
Comment (Atemu) [inline pkgs/by-name/uw/uwsm/package.nix:71]: Why are these deps propagatedBuildInputs?

**Review by Atemu — APPROVED** [review]
This is a clear improvement over the status quo but the PATH wrapping needs to go entirely. 

If UWSM has runtime executable dependencies, their paths should be passed explicitly at configure time or taken from a separate PATH environment variable that is not passed to the compositor etc.

Actually, why does UWSM pass PATH at all? That does not sound sane to me; the PATH for systemd services has nothing to do with what the PATH for the user session.

**Comment by JohnRTitor** [comment]
> If UWSM has runtime executable dependencies, their paths should be passed explicitly at configure time

We already pass python interpreter bin using a meson flag. So wrapping with python at postInstall isn't needed anymore, but the rest still needs to be passed. 

I wonder how to make buildInputs here just work.

File: pkgs/by-name/uw/uwsm/package.nix (hunk 1 above)
Comment (Atemu) [reply pkgs/by-name/uw/uwsm/package.nix:71]: Only python deps need to be propagated and only for packages that other python packages might depend on (this is a leaf package w.r.t. python). There are some other cases where propagating an exact version of some package is important but these are just runtime inputs.

**Comment by Atemu** [comment]
This is something upstream would need to fix in either case.

I think the fact that `PATH` (or any other environment variable besides the global env) is passed to the is the worst offender here and fixing that would make our wrapping okay again.

I'd prefer a more explicit API here; something that would let us pass specific binaries or a lookup path at configure time would be better perhaps.

**Comment by Scrumplex** [comment]
> the PATH for systemd services has nothing to do with what the PATH for the user session.

Isn't that one of the major use cases of UWSM? I mainly rely on UWSM to keep my systemd environment in sync with my user environment, as I start many auxiliary desktop services through systemd

**Comment by Atemu** [comment]
What I meant was the environment that UWSM itself runs in, not the env that it then imports.

**Comment by Vladimir-csp** [comment]
For context, how uwsm's environment preloader works: `sh` (as found by uwsm's python process) is launched with environment derived directly from the current systemd environment. This `sh` sources plugins, shell profile files, and uwsm env files, then prints environment via `exec env -0`. This output is parsed. Difference between initial systemd environment and `env` output (plus/minus some (ex/in)clusions) is then exported into systemd environment.

`PATH` is in always_export set. So it should always result in whatever the final `env` sees, which should come from systemd -> plugins -> profile files -> uwsm env files. Stock plugins do not modify `PATH`.

**Pull request merged**